<div class="title_project">Анализ базы данных сервиса для чтения книг по подписке</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение" data-toc-modified-id="Введение-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение</a></span></li><li><span><a href="#Импорт-библиотек-Python" data-toc-modified-id="Импорт-библиотек-Python-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт библиотек Python</a></span></li><li><span><a href="#Оформление-и-стили" data-toc-modified-id="Оформление-и-стили-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Оформление и стили</a></span></li><li><span><a href="#Подготовка-подключения-к-базе-данных" data-toc-modified-id="Подготовка-подключения-к-базе-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка подключения к базе данных</a></span></li><li><span><a href="#Выполнение-SQL-запросов" data-toc-modified-id="Выполнение-SQL-запросов-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выполнение SQL-запросов</a></span><ul class="toc-item"><li><span><a href="#Изучение-таблиц" data-toc-modified-id="Изучение-таблиц-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Изучение таблиц</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-5.1.4"><span class="toc-item-num">5.1.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-5.1.5"><span class="toc-item-num">5.1.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Расчет-количества-книг,-которые-вышли-после-1-января-2000-года" data-toc-modified-id="Расчет-количества-книг,-которые-вышли-после-1-января-2000-года-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Расчет количества книг, которые вышли после 1 января 2000 года</a></span></li><li><span><a href="#Расчет-количества-обзоров-и-средней-оценки-для-каждой-книги" data-toc-modified-id="Расчет-количества-обзоров-и-средней-оценки-для-каждой-книги-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Расчет количества обзоров и средней оценки для каждой книги</a></span></li><li><span><a href="#Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Определение издательства, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определение-автора-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Определение-автора-с-самой-высокой-средней-оценкой-книг-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Определение автора с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Расчет-среднего-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Расчет-среднего-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>Расчет среднего количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Введение

<b><font color="#4682B4">Цель</font></b> - анализ базы данных сервиса для чтения книг по подписке, чтобы сформулировать ценностное предложение для нового продукта.

---
<b><font color="#4682B4">Задачи:</font></b>
* Посчитать, сколько книг вышло после 1 января 2000 года.
* Для каждой книги посчитать количество обзоров и среднюю оценку.
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц, чтобы исключить из анализа брошюры.
* Определить автора с самой высокой средней оценкой книг. Учитывать только книги с 50 и более оценками.
* Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

---
<b><font color="#4682B4">План исследования:</font></b>
1. Загрузить и изучить данные.
2. Проанализировать полученные данные, выполнив поставленные задачи.
3. Сформулировать выводы.

---
<b><font color="#4682B4">Описание данных:</font></b>  

База данных содержит информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

<b><font color="#4682B4">Таблица books</font></b>  

Содержит данные о книгах:
- `book_id` - идентификатор книги
- `author_id` - идентификатор автора
- `title` - название книги
- `num_pages` - количество страниц
- `publication_date` - дата публикации книги
- `publisher_id` - идентификатор издателя

<b><font color="#4682B4">Таблица authors</font></b>  

Содержит данные об авторах:

- `author_id` - идентификатор автора
- `author` - имя автора

<b><font color="#4682B4">Таблица publishers</font></b>  

Содержит данные об издательствах:

- `publisher_id` - идентификатор издательства
- `publisher` - название издательства

<b><font color="#4682B4">Таблица ratings</font></b>  

Содержит данные о пользовательских оценках книг:

- `rating_id` - идентификатор оценки
- `book_id` - идентификатор книги
- `username` - имя пользователя, оставившего оценку
- `rating` - оценка книги

<b><font color="#4682B4">Таблица reviews</font></b> 

Содержит данные о пользовательских обзорах на книги:

- `review_id` - идентификатор обзора
- `book_id` - идентификатор книги
- `username` - имя пользователя, написавшего обзор
- `text` - текст обзора

## Импорт библиотек Python

In [1]:
# Импорт библиотек
import pandas as pd
from sqlalchemy import create_engine
import os
from dotenv import load_dotenv

In [2]:
# Отключение сообщений об ошибках
import warnings
warnings.filterwarnings('ignore')

## Оформление и стили

In [3]:
# Подключение css
from IPython.core.display import HTML
styles = open('style.css', 'r', encoding='utf-8').read()
display(HTML('<style>%s</style>' % styles))

## Подготовка подключения к базе данных

In [4]:
# Загрузка файла с данными конфигурации
dotenv_path = os.path.join(os.path.dirname(''), '.env')
if os.path.exists(dotenv_path):
    load_dotenv(dotenv_path)

In [5]:
# Храним конфигурацию подключения к БД в отдельной переменной, 
# чтобы её можно было редактировать централизованно или заменить на загрузку и конфигурационного файла
db_config = {'user': os.environ.get('user'), # имя пользователя
             'pwd': os.environ.get('pwd'), # пароль
             'host': os.environ.get('host'),
             'port': os.environ.get('port'), # порт подключения
             'db': os.environ.get('db')} # название базы данных

In [6]:
# Запрашиваем сырые данные
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Выполнение SQL-запросов

### Изучение таблиц

Чтобы изучить таблицы создадим SQL-запросы для каждой из них и выведем первые пять строк. Создадим фунцкцию чтения таблиц:

In [7]:
# Фунцкция чтения таблиц
def SQL(query):
    return pd.io.sql.read_sql(query, con = engine)

#### Таблица books

Изучим таблицу:

In [8]:
SQL("""
       SELECT * FROM books
       LIMIT 5;
    """)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Подсчитаем количество строк в таблице:

In [9]:
SQL("""
       SELECT COUNT(*) 
       FROM books;
    """)

count
0   1000

#### Таблица authors

Изучим таблицу:

In [10]:
SQL("""
       SELECT *
       FROM authors
       LIMIT 5;
    """)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Подсчитаем количество строк в таблице:

In [11]:
SQL("""
       SELECT COUNT(*) 
       FROM authors;
    """)

count
0    636

#### Таблица publishers

Изучим таблицу:

In [12]:
SQL("""
       SELECT *
       FROM publishers
       LIMIT 5;
    """)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Подсчитаем количество строк в таблице:

In [13]:
SQL("""
       SELECT COUNT(*) 
       FROM publishers;
    """)

count
0    340

#### Таблица ratings

Изучим таблицу:

In [14]:
SQL("""
       SELECT *
       FROM ratings
       LIMIT 5;
    """)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Подсчитаем количество строк в таблице:

In [15]:
SQL("""
       SELECT COUNT(*) 
       FROM ratings;
    """)

count
0   6456

#### Таблица reviews



Изучим таблицу:

In [16]:
SQL("""
       SELECT *
       FROM reviews
       LIMIT 5;
    """)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Подсчитаем количество строк в таблице:

In [17]:
SQL("""
       SELECT COUNT(*) 
       FROM reviews;
    """)

count
0   2793

Были загружены и изучены таблицы из базы данных, содержащей информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

**Таблицы:**
1. **books** содержит данные о книгах и состоит из 1000 записей;
2. **authors** содержит данные об авторах и состоит из 636 записей;
3. **publishers** содержит данные об издательствах и состоит из 340 записей;
4. **ratings** содержит данные о пользовательских оценках книг и состоит из 6456 записей;
5. **reviews** содержит данные о пользовательских обзорах на книги и состоит из 2793 записей.

### Расчет количества книг, которые вышли после 1 января 2000 года

Рассчитаем количество книг, которые были опубликованы после 1 января 2000 года:

In [18]:
SQL("""
       SELECT COUNT(book_id) AS cnt_books
       FROM books
       WHERE publication_date > '2000-01-01';
    """)

cnt_books
0        819

В базе найдена 819 книг с датой публикации после 1 января 2000 года.  

Определим есть ли а базе книги, которые были выпущены после 1 января 2000 года в разных вариантах:

In [19]:
SQL("""
       SELECT book_id, publication_date, author, title, num_pages, publisher 
       FROM books 
            INNER JOIN authors ON books.author_id = authors.author_id
            INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
       WHERE title IN (SELECT title 
                       FROM books 
                       GROUP BY title 
                       HAVING COUNT(*) > 1) AND 
                       publication_date >= '2000-01-01';
    """)

book_id publication_date         author                title  num_pages  \
0      426       2005-11-15  Arthur Golden  Memoirs of a Geisha        434   
1      427       2005-11-22  Arthur Golden  Memoirs of a Geisha        503   

                             publisher  
0  Random House Large Print Publishing  
1                    Vintage Books USA

В базе данных 2 книги "Memoirs of a Geisha", изданные после 2000 года встречаются в 2 вариантах издания.

### Расчет количества обзоров и средней оценки для каждой книги

Рассчитаем количество уникальных обзоров и рейтинг для каждой книги:

In [20]:
SQL("""
       SELECT books.book_id,
              authors.author,
              books.title,
              COUNT(DISTINCT reviews.review_id) AS cnt_review,
              AVG(ratings.rating) AS avg_rating
       FROM books 
            LEFT JOIN authors ON books.author_id = authors.author_id
            LEFT JOIN reviews ON books.book_id = reviews.book_id
            LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY books.book_id, authors.author, books.title
        ORDER BY cnt_review DESC
        LIMIT 5;
    """)     

book_id                      author  \
0      948             Stephenie Meyer   
1      497              Diana Gabaldon   
2      207           Elizabeth Gilbert   
3      299  J.K. Rowling/Mary GrandPré   
4      302  J.K. Rowling/Mary GrandPré   

                                               title  cnt_review  avg_rating  
0                            Twilight (Twilight  #1)           7    3.662500  
1                          Outlander (Outlander  #1)           6    4.125000  
2                                    Eat  Pray  Love           6    3.395833  
3  Harry Potter and the Chamber of Secrets (Harry...           6    4.287500  
4  Harry Potter and the Prisoner of Azkaban (Harr...           6    4.414634

В базе присутствуют книги, которые не имеют оценок. Максимальное количество оценок 7 для Twilight #1 Stephenie Meyer. 	

### Определение издательства, которое выпустило наибольшее число книг толще 50 страниц

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц, чтобы исключить из анализа брошюры:

In [21]:
SQL("""
       SELECT publisher,
              COUNT(book_id) AS cnt_book
       FROM publishers
            INNER JOIN books ON books.publisher_id = publishers.publisher_id
       WHERE num_pages > 50
       GROUP BY publisher
       ORDER BY cnt_book DESC
       LIMIT 1;
    """)

publisher  cnt_book
0  Penguin Books        42

Издательство, которое выпустило больше всего книг - Penguin Books. 

### Определение автора с самой высокой средней оценкой книг

Определим автора с самой высокой средней оценкой книг. Учитывать будем только книги с 50 и более оценками:

In [22]:
SQL("""
       SELECT SUBQ.author, ROUND(AVG(SUBQ.avg_rating), 2) AS avg_rating
       FROM (SELECT authors.author,
             AVG(ratings.rating) AS avg_rating
       FROM books 
            LEFT JOIN authors ON books.author_id = authors.author_id
            LEFT JOIN ratings ON books.book_id = ratings.book_id
       GROUP BY books.book_id, authors.author
       HAVING COUNT(DISTINCT ratings.rating_id) > 50) AS SUBQ
       GROUP BY SUBQ.author
       ORDER BY avg_rating DESC
       LIMIT 1;
    """)     

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.28

Автор с самой высокой средней оценкой книг, среди книг с 50 и более оценками - Джоан Роулинг, со средним рейтингом 4,28.

### Расчет среднего количество обзоров от пользователей, которые поставили больше 50 оценок

Рассчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [23]:
SQL("""
       SELECT ROUND(AVG(SUBQ.count), 2) AS avg_cnt_reviews
       FROM (SELECT COUNT(*)
             FROM reviews
             WHERE username IN (SELECT username
                                FROM ratings
                                GROUP BY username
                                HAVING COUNT(rating) > 50)
            GROUP BY username) AS SUBQ;
    """)

avg_cnt_reviews
0            24.33

## Вывод

1. Количество книг в базе данных с датой публикации после 1 января 2000 года - 819.  
2. В базе присутствуют книги, которые не имеют оценок. Максимальное количество оценок 7 для Twilight #1 Stephenie Meyer. 
3. Издательство, которое выпустило наибольшее количество книг - Penguin Books.
4. Автор с самой высокой средней оценкой книг, среди книг с 50 и более оценками - Джоан Роулинг, средний рейтинг 4,28
5. Пользователи, которые поставили больше 50 оценок, оставляют в среднем 24 рецензии.